### PR TITLE
Browse

### PR DESCRIPTION
**Pull request by ttony678**
Add sample browse.json file to populate BrowseFragment
Change ClarityApp and ClarityClient API's to take context as parameter
BrowseFragment GridView should only display toasts when clicked

### CODE REVIEW

**Comment by kkong006** [comment]
Can you please merge with master before requesting a review?

Edit: Merged with master - ready for review.

File: app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java [hunk 1]
```diff
@@ -25,5 +58,81 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+        }
+        return json;
+    }
+
+    private void populateBrowseGrid() {
+        final GridView gridview = getActivity().findViewById(R.id.gv_browse);
+        final ImageTextAdapter adapter = new ImageTextAdapter(getActivity(), channels);
+        gridview.setAdapter(adapter);
+
+        gridview.setOnItemClickListener(new AdapterView.OnItemClickListener() {
```
Comment (kkong006) [inline app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: Can you move this onclicklistener into the ViewHolder class? Please see ChannelFragment.java for an example.

File: app/src/main/java/appricottsoftware/clarity/adapters/ImageTextAdapter.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        ViewHolder holder;
+        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+
+        if (convertView == null) {
+            holder = new ViewHolder();
+            convertView = inflater.inflate(R.layout.item_browse_result, parent, false);
+            holder.textView = convertView.findViewById(R.id.ivBrowseEpisodeName);
+            holder.imageView = convertView.findViewById(R.id.ivBrowseImg);
+            holder.imageView.setPadding(8, 8, 8, 8);
```
Comment (kkong006) [inline app/src/main/java/appricottsoftware/clarity/adapters/ImageTextAdapter.java]: Can you please set the padding attribute in the layout xml?

File: app/src/main/java/appricottsoftware/clarity/adapters/ImageTextAdapter.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    // create a new ImageView for each item referenced by the Adapter
+    public View getView(int position, View convertView, ViewGroup parent) {
+
+        ViewHolder holder;
+        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+
+        if (convertView == null) {
+            holder = new ViewHolder();
+            convertView = inflater.inflate(R.layout.item_browse_result, parent, false);
+            holder.textView = convertView.findViewById(R.id.ivBrowseEpisodeName);
```
Comment (kkong006) [inline app/src/main/java/appricottsoftware/clarity/adapters/ImageTextAdapter.java]: Can you please move the view binding (findviewbyid) into the ViewHolder class? See ChannelFragment.java for how to bind views in a ViewHolder with Butterknife.

File: app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java [hunk 4]
```diff
@@ -25,5 +58,81 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+    }
+
+    // Parses json file and adds contexts to Channel ArrayList
+    void parseJSON() {
+        try {
+            JSONObject obj = new JSONObject(loadJSONFromAsset());
+            JSONArray resp = obj.getJSONArray("results");
+            for (int i = 0; i < resp.length(); i++) {
+                Channel c = getGson().fromJson(String.valueOf(resp.getJSONObject(i)), Channel.class);
+                c.setUid(ClarityApp.getSession(getActivity()).getUserID());
```
Comment (lepatrick714) [inline app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: Can we make this a for each loop?

File: app/src/main/java/appricottsoftware/clarity/sync/ClarityClient.java [hunk 5]
```diff
@@ -76,7 +72,7 @@ public void authenticateUser(String email, String password, JsonHttpResponseHand
 
             JSONObject jsonParams = new JSONObject();
             try {
-                client.setMaxRetriesAndTimeout(1, 1000);
+                client.setMaxRetriesAndTimeout(0, 1000);
```
Comment (lepatrick714) [inline app/src/main/java/appricottsoftware/clarity/sync/ClarityClient.java]: Why did you change it from 1 to 0?

**Review by lepatrick714 — CHANGES_REQUESTED** [review]

File: app/src/main/java/appricottsoftware/clarity/sync/ClarityClient.java (hunk 5 above)
Comment (ttony678) [reply app/src/main/java/appricottsoftware/clarity/sync/ClarityClient.java]: I feel it takes too long to get a toast saying the server is down when there's 2 attempts to reach it. If everyone's content with 1 retry I can put it back (:

File: app/src/main/java/appricottsoftware/clarity/adapters/ImageTextAdapter.java (hunk 2 above)
Comment (ttony678) [reply app/src/main/java/appricottsoftware/clarity/adapters/ImageTextAdapter.java]: Fixed.

**Review by kkong006 — COMMENTED** [review]
Some general stylistic comments on the adapter for the gridview.

File: app/src/main/java/appricottsoftware/clarity/sync/ClarityClient.java (hunk 5 above)
Comment (lepatrick714) [reply app/src/main/java/appricottsoftware/clarity/sync/ClarityClient.java]: Ahhh I understand now!

File: app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java (hunk 4 above)
Comment (ttony678) [reply app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: I don't believe we can. I think the index i is necessary to getting the right JSONobject from the JSONArray variable.

Comment (kkong006) [reply app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: Sorry to hijack this thread, but there's an example of doing it in ChannelFragment and PlayerService. You create a new type (ArrayList<Channel>) and have the gson library deserialize into that type. Something along the lines of:
`TypeToken<ArrayList<Channel>> token = new TypeToken<ArrayList<Channel>>() {};
                    ArrayList<Channel> channels = ClarityApp.getGson().fromJson(response.getObject("results"), token.getType());`

Comment (lepatrick714) [reply app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: Damn it looks like JSONArray uses raw iterators. haha

Comment (lepatrick714) [reply app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: @ttony678 Might be best to follow @kkong006 's templates.

Comment (ttony678) [reply app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: That assumes getting a JSONArray from the parameter. I was under the assumption we'd only be receiving JSONobjects from the  @database?

*EDIT: Ignore me.

Comment (lepatrick714) [reply app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: 💯

Comment (ttony678) [reply app/src/main/java/appricottsoftware/clarity/fragments/BrowseFragment.java]: Using @kkong006 's template. Leaving older code commented because it allows setting UID within the loop which might or might not be useful at this moment.

**Comment by ttony678** [comment]
Currently unable to address @kkong006 's stylistic comments due to using GridView instead of RecyclerView. Using GridView for our current use case since it's much simpler and the total default channels are fixed in size.

**Review by kkong006 — APPROVED** [review]
All comments have been addressed. Looks legit.

**Review by lepatrick714 — APPROVED** [review]
